feat(plugins): add unocss plugin (WIP)

miguelrk opened the pull request:
**TLDR:** [`unocss`](https://unocss.dev/) is an instant on-demand atomic CSS engine and a joy to use. I believe it can (and should) replace `twind` as the default atomic CSS engine for fresh.

### Context

`fresh@v1.2.0` introduced the [`renderAsync`](https://fresh.deno.dev/docs/concepts/plugins#hook-renderasync) method, which allows calling the asynchronous `uno.generate()` method as first [proposed](https://github.com/denoland/fresh/issues/728#issuecomment-1413024482) by @Hladikes (thanks!) but without the workaround. I've tested it for routes, components and islands, and it's working for all (tests are pending). 

### Checklist

This PR is a WIP... I will continue tomorrow with the following:

- [x] enhance defaultConfig
- [x] add tests
- [x] add docs

### Related Issues

https://github.com/denoland/fresh/issues/728
https://github.com/denoland/fresh/issues/290

miguelrk committed:
feat(plugins): add unocss plugin (WIP)

lino-levan commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,29 @@
+`
+
+const defaultConfig: UserConfig = {
+  presets: [presetWind() as unknown as Preset],
```
I wonder if we should use `presetUno()` instead of `presetWind()` here.

miguelrk replied:
I agree, just pushed an update to use `presetUno` instead 👍🏼

miguelrk committed:
fix(plugins/unocss): use presetUno instead of presetWind by default

lino-levan commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,29 @@
+import presetUno from 'https://esm.sh/@unocss/preset-uno@0.46.3?bundle&no-check'
+import { Plugin } from '$fresh/server.ts'
+
+const unoResetCSS = `/* reset */
```
Could you add a comment where you're getting this from?

miguelrk replied:
It's originally from [here](https://github.com/Hladikes/fresh/blob/main/plugins/unocss.ts) but not sure which exact reset is being applied from the [official unocss resets](https://unocss.dev/guide/style-reset). Happy to add a comment...

Hladikes replied:
Not gonna lie, it feels great to have my nickname in comments since this is my first sort of a contribution. However I'm not the original author. Here is the [UnoCSS plugin](https://github.com/ije/html/blob/main/plugins/unocss.ts) by [ije](https://github.com/ije), which I took and slightly modified, and in which he commented the name of the used preset:) Take a look at [here](https://github.com/ije/html/blob/6c5612fb81e5159c1a598eb6b37a3f4c72a25fe5/plugins/unocss.ts#L6C52-L6C52)

miguelrk replied:
Ha! Glad to have you around @Hladikes ! Thanks for the pointer, I will update the comment to point to `https://esm.sh/@unocss/reset@0.51.2/tailwind.css` instead 👍🏼

miguelrk committed:
chore(plugins): add comment to inline reset for unocss

miguelrk committed:
chore(plugins): update comment of inline reset for unocss

miguelrk committed:
Merge branch 'denoland:main' into main

miguelrk committed:
docs(plugins): add docs for unocss

miguelrk commented on docs/latest/examples/using-unocss.md:
````diff
@@ -0,0 +1,60 @@
+await start(manifest, { plugins: [unocssPlugin(unocssConfig)] });
+```
+
+The template generates a Twind v0 project by default. If you want to use UnoCSS
````
I think lines 6-23 are redundant (coming from `using-twind-v1.md`) and could be removed.

lino-levan replied:
Agreed, I think the docs are pretty clear.

miguelrk replied:
Done, removed redundant text in 86c82278775bfe9cbba7d8fb1dcef7b92b247863

lino-levan commented on docs/latest/examples/using-unocss.md:
````diff
@@ -0,0 +1,60 @@
+
+```ts
+import type { UserConfig } from "https://esm.sh/@unocss/core@0.53.1";
+import presetUno from "https://esm.sh/@unocss/preset-uno@0.46.3?bundle&no-check";
````
Is there a reason you're using `?bundle&no-check` here?

miguelrk replied:
Hmm good catch, don't remember where that came from, just pushed a fix 2a64f85735d44e935261995b161d72ab5c549499 removing the `?bundle&no-check`. Tested it and works as well 👍🏼

miguelrk committed:
fix(plugins): remove unnecessary `?bundle&no-check` from `@unocss/preset-uno` import

miguelrk committed:
docs(plugins): remove redundant text for unocss

miguelrk committed:
fix(plugins): remove defaults from unocss plugin

lino-levan commented on docs/latest/examples/using-unocss.md:
````diff
@@ -0,0 +1,42 @@
+
+```ts
+import type { UserConfig } from "https://esm.sh/@unocss/core@0.53.1";
+import presetUno from "https://esm.sh/@unocss/preset-uno@0.46.3";
````
Can we keep all of the versions in this at the same number? It's kind of weird that we have all these different versions throughout the codebase
```suggestion
import type { UserConfig } from "https://esm.sh/@unocss/core@0.54.2";
import presetUno from "https://esm.sh/@unocss/preset-uno@0.54.2";
```

lino-levan commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,25 @@
+import type { UserConfig } from "https://esm.sh/@unocss/core@0.53.1";
+import { UnoGenerator } from "https://esm.sh/@unocss/core@0.53.1";
```
```suggestion
import type { UserConfig } from "https://esm.sh/@unocss/core@0.54.2";
import { UnoGenerator } from "https://esm.sh/@unocss/core@0.54.2";
```
see above

lino-levan commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,25 @@
+import { UnoGenerator } from "https://esm.sh/@unocss/core@0.53.1";
+import { Plugin } from "$fresh/server.ts";
+
+// inline reset from https://esm.sh/@unocss/reset@0.51.2/tailwind.css
```
```suggestion
// inline reset from https://esm.sh/@unocss/reset@0.54.2/tailwind.css
```
see above

lino-levan reviewed (commented):
Outside of tests, this is looking pretty close to ready to me.

miguelrk committed:
chore(plugins): add `fixture_unocss_hydrate` test

miguelrk committed:
chore(plugins): bump `@unocss/reset` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): bump `@unocss/core` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): bump `@unocss/core` and `@unocss/preset-uno` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): format via `deno fmt`

miguelrk committed:
feat(plugins): add `antfu.unocss` to recommended vscode extensions

adamgreg commented on tests/fixture_unocss_hydrate/islands/InsertCssrules.tsx:
I'm sorry if this is a stupid question, but does this work? I can't see how the plugin delivers the unocss JS to the client. Does this just add the `text-green-600` class to the element without defining it in the stylesheet?

I hope review comments from random people aren't too annoying! I just wanted to read through the PR to improve my own understanding, and got stuck on this part.

miguelrk replied:
No prob! There's no such thing as stupid questions 😉 and actually, you just pointed me to an issue with this fixture I think 😆. It appears that the css rule for `text-green-600` is being generated since unocss picks it up from the html "Add \`text-green-600\` to Current Number Class" (=37 css rules in total). If I replace that with e.g. "Add a class to Current Number Class", then unocss won't pick up `text-green-600` from anywhere (=36 css rules in total), so when clicking the button, it won't turn green (since no css rule for `.text-green-600` was generated in the render step).

This plugin only runs on each render (via `renderAsync`), so dynamically updating classes in the client won't cause `unocss` to re-evaluate the html and re-inject css rules (except if you register the [runtime script](https://unocss.dev/integrations/runtime#runtime) via `<script src="https://cdn.jsdelivr.net/npm/@unocss/runtime"></script>`, which might not be desired for this plugin, at least not as a default?). What do you think @lino-levan ? Should we remove this fixture or add an option to inject the unocss runtime script and test it with that?

lino-levan replied:
Yeah, I'm not too sure. Thoughts @marvinhagemeister?

adamgreg replied:
Would it work to just add a boolean option to the plugin config to enable UnoCSS for the client, and add UnoCSS to the `scripts` it returns if the option is set?

Again please excuse my ignorance if I'm missing the whole point. This is how I learn!

miguelrk replied:
Okay, so I've been trying to use some components from `shadcn/ui` and most of them work with the `unocss` plugin, except for dynamically mounted components like dropdowns/dialogs/menus. In these cases, once the dynamic component mounts (e.g. dropdown is activated/opened on click) its html is injected to the DOM client-side (with proper tailwindcss css classes), but these are not picked up by the server-side `pluginUnocss`. Adding the `unocss` runtime script fixes this, since styles are generated client-side dynamically whenever DOM changes are detected. 

Taking that into account, I now think that this `pluginUnocss` should include the client-side runtime by default (to avoid headaches) or at least have the option to include it. The good thing is, that since its a mere javascript engine, its lightweight at **167.61KB** for `https://cdn.jsdelivr.net/npm/@unocss/runtime`.

I hope this discovery sheds some light. How does twind handle this if at all?

lino-levan replied:
I believe twind ships the engine to the client side. That being said, I'm in favor of including the engine by default, with an option to disable it.

miguelrk replied:
Just pushed a7243b15e1bf98b4e6bf5ade503adda5dc9209ce to include the `unocss` runtime script by default (I agree this is best), but with the option to disable it. The plugin signature is now 

```ts
function unocss(config: UserConfig, runtime = true): Plugin
```

With this changes, the `insert-cssrules` fixture test is now working properly:

![Screenshot 2023-08-10 8 07 55 AM](https://github.com/denoland/fresh/assets/18708856/f77c8f7e-300f-4d0a-9ce6-a9625a480496)

One last issue is [preventing FOUC](https://unocss.dev/integrations/runtime#preventing-fouc). For smaller components like that of this test, there is no perceivable "flash of unstyled content", however, with larger components like the `shadcn/ui` dropdown, this flash is very apparent. 

That being said, since this requires users to manually add a custom selector e.g. an `un-cloak` attribute and a CSS rule `[un-cloak] { display: none; }` to circumvent this (as suggested by unocss docs), I think this is out of scope for this plugin. Open to further discuss/improve this.

miguelrk committed:
chore(ci): try fixing broken types causing CI to fail

miguelrk committed:
chore(plugins): order imports

miguelrk committed:
chore(plugins): use import map instead of url imports for unocss

miguelrk committed:
feat(plugins): add client runtime script to unocss plugin by default

miguelrk committed:
chore(plugins): unify imports form `@unocss/core`

adamgreg committed:
Re-export UserConfig type from UnoCSS plugin

adamgreg committed:
Import @unocss/preset-uno from full URL in test

adamgreg committed:
Update UnoCSS in plugin to 0.55.1

adamgreg committed:
In the UnoCSS plugin pass the config to runtime

-  Define a Config interface for the plugin, which extends UnoCSS UserConfig with a selfURL property
- Add a `defineConfig()` function for help with types, and make it easier to adapt existing UnoCSS config modules for use with the plugin
- Move the runtime script inline

miguelrk committed:
Merge pull request #1 from adamgreg/main

Rework UnoCSS plugin imports

miguelrk committed:
chore(plugins): format via `deno fmt`

miguelrk committed:
Merge branch 'main' into main

zhmushan commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,45 @@
+          window.__unocss = config;
+          const script = document.createElement("script");
+          script.src = "https://cdn.jsdelivr.net/npm/@unocss/runtime";
+          document.head.appendChild(script);
```
```suggestion
        import init from "https://esm.sh/@unocss/runtime@0.55.1";
        export default function() {
          window.__unocss = config;
          init();
```

zhmushan commented on docs/latest/examples/using-unocss.md:
```diff
@@ -0,0 +1,43 @@
+
+export default {
+  presets: [presetUno()],
+} satisfies UserConfig;
```
```suggestion
import { defineConfig } from "$fresh/plugins/unocss.ts";
import presetUno from "https://esm.sh/@unocss/preset-uno@0.55.1";

export default defineConfig({
  presets: [presetUno()],
  selfURL: import.meta.url,
});
```

miguelrk committed:
Update plugins/unocss.ts

Use `init()` function of `@unocss/runtime` instead of manually injecting the script tag to avoid fetching it (redundantly) in the client, and instead ship runtime with `plugin-unocss.main.ts`.

Co-authored-by: 木杉 <zhmushan@qq.com>

miguelrk committed:
chore(plugins): update querySelector to style[data-unocss-runtime-layer="default"] in fixture_unocss_hydrate

miguelrk committed:
Merge branch 'main' of https://github.com/miguelrk/fresh

miguelrk committed:
Update docs/latest/examples/using-unocss.md

Co-authored-by: 木杉 <zhmushan@qq.com>

adamgreg committed:
UnoCSS plugin use uno.config.ts

- Import config from uno.config.ts if no config object is explicitly provided
- Always use uno.config.ts as the import source for the browser runtime config
- This avoids the complexity of selfURL, allowing the plugin to use standard Uno config files

adamgreg committed:
UnoCSS plugin: Add defineConfig() helper function

It is the same as found in the unocss package, but that can not currently be safely imported in Deno (due to Node-specific code in the icons preset).

miguelrk committed:
Merge pull request #2 from adamgreg/config_file

UnoCSS plugin use uno.config.ts

miguelrk committed:
Merge branch 'denoland:main' into main

adamgreg committed:
Make UnoCSS plugin function synchronous

miguelrk committed:
Merge pull request #3 from adamgreg/config_file

Make UnoCSS plugin function synchronous

miguelrk committed:
Merge branch 'main' into main

adamgreg committed:
UnoCSS Plugin: Use Preact hook for SSR

miguelrk committed:
Merge pull request #4 from adamgreg/main

UnoCSS: Use Preact hook for SSR
